mama/c_cpp/src/c/publisher.c

```diff
@@ -611,3 +719,280 @@ mamaPublisher_getTransport (mamaPublisher   publisher,
+            /* Not in destroy() finish */
+            mamaPublisherImpl_setState(impl, MAMA_PUBLISHER_DESTROYED);
+            wlock_unlock(impl->mCreateDestroyLock);
+            mamaPublisherImpl_cleanup(impl);
+            return;
+            break;
+        default:
+            /* TODO */
+            break;
+    }
+    wlock_unlock(impl->mCreateDestroyLock);
+}
+
+static void mamaPublisherImpl_setState(mamaPublisherImpl* impl, mamaPublisherState state)
+{
+    /* Set the state using an atomic operation so it will be thread safe. */
+    wInterlocked_set(state, (wInterlockedInt*) &impl->mState);
+    if (mama_getLogLevel() >= MAMA_LOG_LEVEL_FINEST)
+    {
+        const char* source = "(none)";
+        const char* symbol = "(none)";;
+        mamaPublisher_getSource((mamaPublisher) impl, &source);
+        mamaPublisher_getSymbol((mamaPublisher) impl, &symbol);
+        mama_log(MAMA_LOG_LEVEL_FINEST, "Publisher %p %s.%s is now at state %s.",
+            impl, source, symbol, mamaPublisher_stringForState(state));
+    }
+}
+
+static mama_status mamaPublisherImpl_destroy(mamaPublisherImpl *impl)
+{
+    /* Returns. */
+    mama_status        status   = MAMA_STATUS_OK;
+
+    wlock_lock(impl->mCreateDestroyLock);
+
+    /* Set the state to be deleted to at least show in the log that it has been completely removed. */
+    mamaPublisherImpl_setState(impl, MAMA_PUBLISHER_DESTROYING);
+                
+    mamaTransport_removePublisher(impl->mTport, impl->mTransportHandle);
+
+    if (!impl->mMamaPublisherBridgeImpl)
+    {
+        mama_log (MAMA_LOG_LEVEL_ERROR,
+                  "mamaPublisher_destroy() No publisher impl. Cannot destroy.");
+        status = MAMA_STATUS_INVALID_ARG;
+    }
+
+    if (!impl->mBridgeImpl)
+    {
+        mama_log (MAMA_LOG_LEVEL_ERROR,
+                  "mamaPublisher_destroy() No bridge impl. Cannot destroy.");
+        status = MAMA_STATUS_INVALID_ARG;
+    }
+
+    if (impl->mPendingActions && list_size (impl->mPendingActions))
+    {
+        if (MAMA_STATUS_OK!=(status=mamaTransport_throttleRemoveFromList (
+                                                impl->mTport,
+                                                MAMA_THROTTLE_DEFAULT,
+                                                impl->mPendingActions)))
+        {
+            mama_log (MAMA_LOG_LEVEL_ERROR, "mamaPublisher_destroy():"
+                    " Failed to remove actions for publisher");
+        }
+    }
+
+    list_destroy (impl->mPendingActions, NULL, NULL);
+
+    wlock_unlock(impl->mCreateDestroyLock);
+
+    if (impl->mBridgeImpl &&
+            impl->mMamaPublisherBridgeImpl)
+    {
+        if (MAMA_STATUS_OK!=(status=
+                    impl->mBridgeImpl->bridgeMamaPublisherDestroy
+                    (impl->mMamaPublisherBridgeImpl)))
+        {
+           mama_log (MAMA_LOG_LEVEL_ERROR, "mamaPublisher_destroy(): "
+                     "Could not destroy publisher bridge.");
```

Comment:
If you're going to assign the status here, you're best using it in the log message too with a stringForStatus call.